File: src/process/resources/skills/company-analyzer/scripts/read-docx.py
"""Read a Word document and output text content.
Supports: .docx, .doc, .rtf, .odt
Usage: python read-docx.py <file_path>
"""
import sys, os, subprocess
if len(sys.argv) < 2: print("Usage: python read-docx.py <file_path>"); sys.exit(1)
path = sys.argv[1]
ext = os.path.splitext(path)[1].lower()

if ext == '.docx':
    from docx import Document
    doc = Document(path)
    for para in doc.paragraphs:
        if para.text.strip():
            print(para.text)
    for i, table in enumerate(doc.tables):
        print(f"\n[Table {i+1}]")
        for row in table.rows:
            print("\t".join(cell.text.strip() for cell in row.cells))
elif ext == '.doc':
    # Try antiword first, then LibreOffice conversion
    try:
        result = subprocess.run(['antiword', path], capture_output=True, text=True, timeout=30)
        if result.returncode == 0:
            print(result.stdout)
        else:
            raise Exception("antiword failed")
    except Exception:
        # Fallback: use python-docx2txt or textract if available
        try:
            import docx2txt
            print(docx2txt.process(path))
        except ImportError:
            # Last resort: try reading as raw text
            try:
                with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
                    # Filter out binary garbage, keep printable text
                    clean = ''.join(c for c in content if c.isprintable() or c in '\n\r\t')
                    print(clean)
            except Exception as e:
                print(f"Cannot read .doc file: {e}")
                print("Install python-docx2txt: pip install docx2txt")
                sys.exit(1)
elif ext == '.rtf':
    try:
        from striprtf.striprtf import rtf_to_text
        with open(path, 'r', errors='ignore') as f:
            print(rtf_to_text(f.read()))
    except ImportError:
        with open(path, 'r', errors='ignore') as f:
            print(f.read())
else:
    # .odt and others — try reading as text
    with open(path, 'r', encoding='utf-8', errors='ignore') as f:
        print(f.read())

File: src/process/resources/skills/company-analyzer/scripts/read-docx.js
#!/usr/bin/env node
/**
 * Read a Word document and output text content.
 * Supports: .docx, .doc
 * Uses PowerShell ZIP/XML extraction (no npm dependencies).
 * Usage: node read-docx.js <file_path>
 */
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

if (process.argv.length < 3) { console.log('Usage: node read-docx.js <file_path>'); process.exit(1); }

const filePath = path.resolve(process.argv[2]);

// Try Python first (best quality)
const pyScript = path.join(__dirname, 'read-docx.py');
if (fs.existsSync(pyScript)) {
  try {
    const result = execSync(`python "${pyScript}" "${filePath}"`, {
      encoding: 'utf-8', timeout: 30000, maxBuffer: 10 * 1024 * 1024, stdio: ['pipe', 'pipe', 'pipe'],
    });
    if (result.trim()) { console.log(result); process.exit(0); }
  } catch (_) { /* fall through */ }
}

// Fallback: DOCX is a ZIP — extract text from word/document.xml
try {
  const escapedPath = filePath.replace(/'/g, "''");
  const cmd = `powershell -Command "Add-Type -A 'System.IO.Compression.FileSystem'; $z = [IO.Compression.ZipFile]::OpenRead('${escapedPath}'); $doc = $z.Entries | Where-Object { $_.FullName -eq 'word/document.xml' } | Select-Object -First 1; if ($doc) { $r = New-Object IO.StreamReader($doc.Open()); $xml = $r.ReadToEnd(); $r.Close(); $xml = $xml -replace '<w:br[^/]*/>', [char]10; $xml = $xml -replace '</w:p>', ([char]10); $xml = $xml -replace '<[^>]+>', ''; $xml = $xml -replace '&amp;', '&'; $xml = $xml -replace '&lt;', '<'; $xml = $xml -replace '&gt;', '>'; $xml = $xml -replace '&quot;', '\"'; Write-Output $xml.Trim() } else { Write-Output '[No document content found]' }; $z.Dispose()"`;
  const result = execSync(cmd, { encoding: 'utf-8', timeout: 15000, maxBuffer: 10 * 1024 * 1024 });
  console.log(result.trim() ? result : '[No text content found]');
} catch (e) {
  console.error(`Error reading ${filePath}: ${e.message}`);
  process.exit(1);
}
